Userland/Libraries/LibWeb/Layout/GridFormattingContext.h

```diff
@@ -16,22 +16,30 @@ enum class GridDimension {
     Column
 };
 
+struct GridPosition {
+    size_t row;
+    size_t column;
+    inline bool operator==(GridPosition const&) const = default;
+};
+
 class OccupationGrid {
 public:
-    OccupationGrid(size_t column_count, size_t row_count);
-    OccupationGrid();
+    OccupationGrid(size_t columns_count, size_t rows_count)
+        : m_columns_count(columns_count)
+        , m_rows_count(rows_count) {};
+    OccupationGrid() {};
 
-    void maybe_add_column(size_t needed_number_of_columns);
-    void maybe_add_row(size_t needed_number_of_rows);
     void set_occupied(size_t column_start, size_t column_end, size_t row_start, size_t row_end);
     void set_occupied(size_t column_index, size_t row_index);
 
-    size_t column_count() { return m_occupation_grid[0].size(); }
-    size_t row_count() { return m_occupation_grid.size(); }
-    bool is_occupied(size_t column_index, size_t row_index);
+    size_t column_count() const { return m_columns_count; }
+    size_t row_count() const { return m_rows_count; }
+    bool is_occupied(size_t column_index, size_t row_index) const;
 
 private:
-    Vector<Vector<bool>> m_occupation_grid;
+    HashMap<GridPosition, bool> m_occupation_grid;
```

Comment:
Can we make this a simple bitmap instead? Or even just `FixedArray<bool>`?
(And then address into it via `m_occupation_grid[row * column_count + column]`)

Since we know exactly how many elements we need, using a `HashMap` seems unnecessarily expensive (rehashing, reallocations, etc) compared to a simple boolean array.

Reply:
although OccupationGrid constructor accepts number of columns and rows these might not be the final sizes of the grid. that is why there were maybe_add_row/column calls to resize underlying `Vector<Vector<bool>>` in case "implicit" row or column is required during positioning. It is going to be possible to use `Vector<bool>` and address by `m_occupation_grid[row * column_count + column]` but then OccupationGrid would have to care about resizing which is not the case with HashMap.